pkg/controller/direct/redis/cluster/rediscluster_controller.go

```diff
@@ -0,0 +1,284 @@
+	if err := monitoring.VisitFields(obj, &refNormalizer{ctx: ctx, src: obj, project: *projectRef, kube: kube}); err != nil {
+		return nil, err
+	}
+
+	mapCtx := &direct.MapContext{}
+	desiredProto := RedisClusterSpec_ToProto(mapCtx, &obj.Spec)
+	if mapCtx.Err() != nil {
+		return nil, mapCtx.Err()
+	}
+
+	return &redisClusterAdapter{
+		projectID:      projectID,
+		location:       location,
+		resourceID:     resourceID,
+		desired:        desiredProto,
+		clustersClient: redisClustersClient,
+	}, nil
+}
+
+func (m *redisClusterModel) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	return nil, nil
+}
+
+// Find implements the Adapter interface.
+func (a *redisClusterAdapter) Find(ctx context.Context) (bool, error) {
+	if a.resourceID == "" {
+		return false, nil
+	}
+
+	req := &pb.GetClusterRequest{
+		Name: a.fullyQualifiedName(),
+	}
+	redisCluster, err := a.clustersClient.GetCluster(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	a.actual = redisCluster
+
+	return true, nil
+}
+
+// Delete implements the Adapter interface.
+func (a *redisClusterAdapter) Delete(ctx context.Context) (bool, error) {
+	// Already deleted
+	if a.resourceID == "" {
+		return false, nil
+	}
+
+	// TODO: Delete via status selfLink?
+	req := &pb.DeleteClusterRequest{
+		Name: a.fullyQualifiedName(),
+	}
+
+	op, err := a.clustersClient.DeleteCluster(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("deleting redisCluster %s: %w", a.fullyQualifiedName(), err)
+	}
+
+	if err := op.Wait(ctx); err != nil {
+		return false, fmt.Errorf("waiting for redisCluster delete %s: %w", a.fullyQualifiedName(), err)
+	}
+
+	return true, nil
+}
+
+func (a *redisClusterAdapter) Export(ctx context.Context) (*unstructured.Unstructured, error) {
+	return nil, nil
+}
+
+// Create implements the Adapter interface.
+func (a *redisClusterAdapter) Create(ctx context.Context, u *unstructured.Unstructured) error {
+	log := klog.FromContext(ctx)
+	log.V(0).Info("creating object", "u", u)
+
+	parent := "projects/" + a.projectID + "/locations/" + a.location
+
+	req := &pb.CreateClusterRequest{
+		Parent:    parent,
+		ClusterId: a.resourceID,
+		Cluster:   a.desired,
+	}
+	log.V(0).Info("making redis CreateCluster call", "request", req)
+
+	op, err := a.clustersClient.CreateCluster(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating cluster: %w", err)
+	}
+
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("waiting for redisCluster create %s: %w", a.fullyQualifiedName(), err)
+	}
+
+	log.V(0).Info("created redisCluster", "redisCluster", created)
+
+	resourceID := lastComponent(created.Name)
+	if err := unstructured.SetNestedField(u.Object, resourceID, "spec", "resourceID"); err != nil {
+		return fmt.Errorf("setting spec.resourceID: %w", err)
+	}
```

Comment:
Is this write back to spec? I think RedisCluster is alpha resource and we don't need to write it back or maybe I mis-understand this.

Reply:
This is writing back to spec, yes.  Remember this code was originally written 3 months ago ... basically prehistoric.

I don't think we've written down (on GitHub?) that we don't want to write back to spec.resourceID for new SciFi resources.  I'd be in favor of that cc @cheftako

Reply:
I changed it not to write back to spec!